http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutor.java

```diff
@@ -124,23 +207,134 @@ private void executeRequest(CrtRequestContext executionContext,
                 // it's semantically an IOException anyway.
                 toThrow = new IOException(e);
             }
-            reportFailure(crtConn,
+            reportAsyncFailure(crtConn,
                           toThrow,
                           requestFuture,
                           asyncRequest.responseHandler());
         } catch (IllegalStateException | CrtRuntimeException e) {
             // CRT throws IllegalStateException if the connection is closed
-            reportFailure(crtConn, new IOException("An exception occurred when making the request", e),
+            reportAsyncFailure(crtConn, new IOException("An exception occurred when making the request", e),
                           requestFuture,
                           asyncRequest.responseHandler());
         }
     }
 
+    private static final class InputStreamAdaptingHttpStreamResponseHandler implements HttpStreamResponseHandler {
+        private final SdkHttpFullResponse.Builder responseBuilder = SdkHttpFullResponse.builder();
+        private volatile InputStreamSubscriber inputStreamSubscriber;
+        private volatile SimplePublisher<ByteBuffer> simplePublisher;
+
+        private final CompletableFuture<SdkHttpFullResponse> requestCompletionFuture;
+        private final HttpClientConnection crtConn;
+
+        InputStreamAdaptingHttpStreamResponseHandler(HttpClientConnection crtConn,
+                                                     CompletableFuture<SdkHttpFullResponse> requestCompletionFuture) {
+            this.crtConn = crtConn;
+            this.requestCompletionFuture = requestCompletionFuture;
+        }
+
+        @Override
+        public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType,
+                                      HttpHeader[] nextHeaders) {
+            for (HttpHeader header : nextHeaders) {
+                responseBuilder.putHeader(header.getName(), header.getValue());
+            }
+
+            if (responseBuilder.statusCode() == 0) {
+                responseBuilder.statusCode(responseStatusCode);
+            }
+
+        }
+
+        @Override
+        public void onResponseComplete(HttpStream stream, int errorCode) {
+            // always close the connection on a 5XX response code.
+            if (HttpStatusFamily.of(stream.getResponseStatusCode()) == HttpStatusFamily.SERVER_ERROR) {
+                crtConn.shutdown();
+            }
+
+            if (errorCode == 0) {
+                requestCompletionFuture.complete(responseBuilder.build());
+            } else {
+                requestCompletionFuture.completeExceptionally(new HttpException(errorCode));
```

Comment:
Should we check the error code and complete the future with IOException here if it's a retryable error code?